Fix error handling of wrapped errors

fabiante opened the pull request:
This PR fixes error handling for wrapped errors.

This bug was caused by wrong parameter order in the usage of `errors.Is`

The added test case shows a use case.

@richzw This changes the behaviour of the middleware: Now - errors which may deeply wrap other errors could match in the middleware where they previously didn't. I think this change fixes a bug but maybe you want the behaviour do not change? LMK.

fabiante committed:
Add test case for wrapped error

This exposes a bug in the existing implementation.
Wrapped errors are not handled by the middleware.

fabiante committed:
Fix error comparison using errors.Is

The parameters used where in the wrong order.

richzw reviewed (changes requested):
Thank you very much for your PR. I think it could be better for `if errors.Is(lastError.Err, e) || errors.Is(e, lastError.Err) {` since the only the first parameter of `errors.is` was `Unwrap` .

fabiante committed:
Refactor error matching into dedicated function

This allows simpler unit tests of the matching logic and reduces complexity
of the original function.

fabiante committed:
Add unit tests on new match function

These ensure the matching will stay the same for future changes
without the testing overhead of performing an http request.

fabiante committed:
Switch comparisons in error matching

This has been suggested in PR review.

fabiante committed:
Remove redundant error matching.

It seems that all tests are fine without the second comparison.

fabiante commented:
Thx for the feedback. I think it is useful to move the error matching into it's own function which allowed for far simpler unit testing.

After I applied you change, which I don't yet fully get, I wanted to test if we even need two comparisons. Have a look at the changes and tell me what you think. I think only one comparison is actually needed.

Tests are green.

richzw reviewed (approved):
LGTM. Thank you very much for your PR
